standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreChecker.java

```diff
@@ -326,10 +352,24 @@ void checkTable(Table table, PartitionIterable parts, byte[] filterExp, CheckRes
       CheckResult.PartitionResult prFromMetastore = new CheckResult.PartitionResult();
       prFromMetastore.setPartitionName(getPartitionName(table, partition));
       prFromMetastore.setTableName(partition.getTableName());
-      if (!fs.exists(partPath)) {
-        result.getPartitionsNotOnFs().add(prFromMetastore);
-      } else {
+      if (allPartDirs.remove(partPath)) {
         result.getCorrectPartitions().add(prFromMetastore);
+      } else {
+        // There can be edge case where user can define partition directory outside of table directory
+        // to avoid eviction of such partitions
+        // we check existence of partition path which are not in table directory
+        if (!partPath.toString().contains(tablePathStr)) {
+          if (!fs.exists(partPath)) {
+            result.getPartitionsNotOnFs().add(prFromMetastore);
+          } else {
+            result.getCorrectPartitions().add(prFromMetastore);
+          }
+        } else {
+          // If Partition Path contains table path, we assume to be non-existent partition since
```

Comment:
Is this a valid assumption? I mean say if the custom partition is inside the table directory but doesn't match the HIVE_DEFAIULT_PARTITON i.e the custom partition points to any folder inside the table directory shouldn't we check its existence in file system as well?

Reply:
It regress the existing behavior where we check fs.exists() for all the partition paths from the metaatsore and add them to  result.getPartitionsNotOnFs().add(prFromMetastore); as requried.